File: objectiveai-js/src/chat/completions/response/usage.ts
import { numberIsEmpty } from "src/isEmpty";
import z from "zod";
import { convert, type JSONSchema } from "../../../json_schema";

export const UsageCompletionTokensDetailsSchema = z
  .object({
    accepted_prediction_tokens: z
      .uint32()
      .optional()
      .describe("The number of accepted prediction tokens in the completion."),
    audio_tokens: z
      .uint32()
      .optional()
      .describe("The number of generated audio tokens in the completion."),
    reasoning_tokens: z
      .uint32()
      .optional()
      .describe("The number of generated reasoning tokens in the completion."),
    rejected_prediction_tokens: z
      .uint32()
      .optional()
      .describe("The number of rejected prediction tokens in the completion."),
  })
  .describe("Detailed breakdown of generated completion tokens.");
export type UsageCompletionTokensDetails = z.infer<
  typeof UsageCompletionTokensDetailsSchema
>;
export const UsageCompletionTokensDetailsJsonSchema: JSONSchema = convert(
  UsageCompletionTokensDetailsSchema,
);

export namespace UsageCompletionTokensDetails {
  export function isEmpty({
    accepted_prediction_tokens,
    audio_tokens,
    reasoning_tokens,
    rejected_prediction_tokens,
  }: UsageCompletionTokensDetails): boolean {
    return (
      numberIsEmpty(accepted_prediction_tokens) &&
      numberIsEmpty(audio_tokens) &&
      numberIsEmpty(reasoning_tokens) &&
      numberIsEmpty(rejected_prediction_tokens)
    );
  }
}

export const UsagePromptTokensDetailsSchema = z
  .object({
    audio_tokens: z
      .uint32()
      .optional()
      .describe("The number of audio tokens in the prompt."),
    cached_tokens: z
      .uint32()
      .optional()
      .describe("The number of cached tokens in the prompt."),
    cache_write_tokens: z
      .uint32()
      .optional()
      .describe("The number of prompt tokens written to cache."),
    video_tokens: z
      .uint32()
      .optional()
      .describe("The number of video tokens in the prompt."),
  })
  .describe("Detailed breakdown of prompt tokens.");
export type UsagePromptTokensDetails = z.infer<
  typeof UsagePromptTokensDetailsSchema
>;
export const UsagePromptTokensDetailsJsonSchema: JSONSchema = convert(
  UsagePromptTokensDetailsSchema,
);

export namespace UsagePromptTokensDetails {
  export function isEmpty({
    audio_tokens,
    cached_tokens,
    cache_write_tokens,
    video_tokens,
  }: UsagePromptTokensDetails): boolean {
    return (
      numberIsEmpty(audio_tokens) &&
      numberIsEmpty(cached_tokens) &&
      numberIsEmpty(cache_write_tokens) &&
      numberIsEmpty(video_tokens)
    );
  }
}

export const UsageCostDetailsSchema = z
  .object({
    upstream_inference_cost: z
      .number()
      .optional()
      .describe("The cost incurred upstream."),
    upstream_upstream_inference_cost: z
      .number()
      .optional()
      .describe("The cost incurred by upstream's upstream."),
  })
  .describe("Detailed breakdown of upstream costs incurred.");
export type UsageCostDetails = z.infer<typeof UsageCostDetailsSchema>;
export const UsageCostDetailsJsonSchema: JSONSchema = convert(
  UsageCostDetailsSchema,
);

export namespace UsageCostDetails {
  export function isEmpty({
    upstream_inference_cost,
    upstream_upstream_inference_cost,
  }: UsageCostDetails): boolean {
    return (
      numberIsEmpty(upstream_inference_cost) &&
      numberIsEmpty(upstream_upstream_inference_cost)
    );
  }
}

export const UsageSchema = z
  .object({
    completion_tokens: z
      .uint32()
      .describe("The number of tokens generated in the completion."),
    prompt_tokens: z.uint32().describe("The number of tokens in the prompt."),
    total_tokens: z
      .uint32()
      .describe(
        "The total number of tokens used in the prompt or generated in the completion.",
      ),
    completion_tokens_details: UsageCompletionTokensDetailsSchema.optional(),
    prompt_tokens_details: UsagePromptTokensDetailsSchema.optional(),
    cost: z
      .number()
      .describe("The cost in credits incurred for this completion."),
    cost_details: UsageCostDetailsSchema.optional(),
    total_cost: z
      .number()
      .describe("The total cost in credits incurred including upstream costs."),
    cost_multiplier: z
      .number()
      .describe(
        "The cost multiplier applied to upstream costs for computing ObjectiveAI costs.",
      ),
    is_byok: z
      .boolean()
      .describe(
        "Whether the completion used a BYOK (Bring Your Own Key) API Key.",
      ),
  })
  .describe("Token and cost usage statistics for the completion.");
export type Usage = z.infer<typeof UsageSchema>;
export const UsageJsonSchema: JSONSchema = convert(UsageSchema);

export namespace Usage {
  export function isEmpty({
    completion_tokens,
    prompt_tokens,
    total_tokens,
    completion_tokens_details,
    prompt_tokens_details,
    cost,
    cost_details,
    total_cost,
  }: Usage): boolean {
    return (
      numberIsEmpty(completion_tokens) &&
      numberIsEmpty(prompt_tokens) &&
      numberIsEmpty(total_tokens) &&
      (completion_tokens_details === undefined ||
        UsageCompletionTokensDetails.isEmpty(completion_tokens_details)) &&
      (prompt_tokens_details === undefined ||
        UsagePromptTokensDetails.isEmpty(prompt_tokens_details)) &&
      numberIsEmpty(cost) &&
      (cost_details === undefined || UsageCostDetails.isEmpty(cost_details)) &&
      numberIsEmpty(total_cost)
    );
  }
}

File: objectiveai-rs/src/chat/completions/response/usage.rs
//! Token usage and cost information.

use crate::chat::completions::response::util;
use serde::{Deserialize, Serialize};

/// Token usage and cost statistics for a completion.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Usage {
    /// Number of tokens in the completion.
    pub completion_tokens: u64,
    /// Number of tokens in the prompt.
    pub prompt_tokens: u64,
    /// Total tokens (prompt + completion).
    pub total_tokens: u64,
    /// Detailed breakdown of completion tokens.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completion_tokens_details: Option<CompletionTokensDetails>,
    /// Detailed breakdown of prompt tokens.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_tokens_details: Option<PromptTokensDetails>,
    /// The cost charged by ObjectiveAI for this request.
    pub cost: rust_decimal::Decimal,
    /// Detailed cost breakdown.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost_details: Option<CostDetails>,
    /// Total cost including ObjectiveAI's charge plus all upstream charges.
    /// For BYOK requests, ObjectiveAI only charges the cost_multiplier difference,
    /// but total_cost still includes what the upstream provider charged.
    pub total_cost: rust_decimal::Decimal,
    /// The multiplier applied to compute ObjectiveAI's charge.
    pub cost_multiplier: rust_decimal::Decimal,
    /// Whether this request used Bring Your Own Key (BYOK).
    pub is_byok: bool,
}

impl Usage {
    /// Appends usage statistics from another instance.
    pub fn push(&mut self, other: &Usage) {
        self.completion_tokens += other.completion_tokens;
        self.prompt_tokens += other.prompt_tokens;
        self.total_tokens += other.total_tokens;
        match (
            &mut self.completion_tokens_details,
            &other.completion_tokens_details,
        ) {
            (Some(self_details), Some(other_details)) => {
                self_details.push(other_details);
            }
            (None, Some(other_details)) => {
                self.completion_tokens_details = Some(other_details.clone());
            }
            _ => {}
        }
        match (
            &mut self.prompt_tokens_details,
            &other.prompt_tokens_details,
        ) {
            (Some(self_details), Some(other_details)) => {
                self_details.push(other_details);
            }
            (None, Some(other_details)) => {
                self.prompt_tokens_details = Some(other_details.clone());
            }
            _ => {}
        }
        self.cost += other.cost;
        match (&mut self.cost_details, &other.cost_details) {
            (Some(self_cost_details), Some(other_cost_details)) => {
                self_cost_details.push(other_cost_details);
            }
            (None, Some(other_cost_details)) => {
                self.cost_details = Some(other_cost_details.clone());
            }
            _ => {}
        }
        self.total_cost += other.total_cost;
    }
}

/// Detailed breakdown of completion token usage.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct CompletionTokensDetails {
    /// Tokens from accepted predictions (speculative decoding).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accepted_prediction_tokens: Option<u64>,
    /// Audio output tokens.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audio_tokens: Option<u64>,
    /// Tokens used for reasoning/thinking.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_tokens: Option<u64>,
    /// Tokens from rejected predictions (speculative decoding).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rejected_prediction_tokens: Option<u64>,
}

impl CompletionTokensDetails {
    /// Returns `true` if any token count is non-zero.
    pub fn any_usage(&self) -> bool {
        self.accepted_prediction_tokens.is_some_and(|v| v > 0)
            || self.audio_tokens.is_some_and(|v| v > 0)
            || self.reasoning_tokens.is_some_and(|v| v > 0)
            || self.rejected_prediction_tokens.is_some_and(|v| v > 0)
    }

    /// Appends token details from another instance.
    pub fn push(&mut self, other: &CompletionTokensDetails) {
        util::push_option_u64(
            &mut self.accepted_prediction_tokens,
            &other.accepted_prediction_tokens,
        );
        util::push_option_u64(&mut self.audio_tokens, &other.audio_tokens);
        util::push_option_u64(
            &mut self.reasoning_tokens,
            &other.reasoning_tokens,
        );
        util::push_option_u64(
            &mut self.rejected_prediction_tokens,
            &other.rejected_prediction_tokens,
        );
    }
}

/// Detailed breakdown of prompt token usage.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct PromptTokensDetails {
    /// Audio input tokens.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audio_tokens: Option<u64>,
    /// Tokens served from cache.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cached_tokens: Option<u64>,
    /// Tokens written to cache.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cache_write_tokens: Option<u64>,
    /// Video input tokens.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub video_tokens: Option<u64>,
}

impl PromptTokensDetails {
    /// Returns `true` if any token count is non-zero.
    pub fn any_usage(&self) -> bool {
        self.audio_tokens.is_some_and(|v| v > 0)
            || self.cached_tokens.is_some_and(|v| v > 0)
            || self.cache_write_tokens.is_some_and(|v| v > 0)
            || self.video_tokens.is_some_and(|v| v > 0)
    }

    /// Appends token details from another instance.
    pub fn push(&mut self, other: &PromptTokensDetails) {
        util::push_option_u64(&mut self.audio_tokens, &other.audio_tokens);
        util::push_option_u64(&mut self.cached_tokens, &other.cached_tokens);
        util::push_option_u64(
            &mut self.cache_write_tokens,
            &other.cache_write_tokens,
        );
        util::push_option_u64(&mut self.video_tokens, &other.video_tokens);
    }
}

/// Detailed cost breakdown.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostDetails {
    /// Cost charged by the immediate upstream (e.g., OpenRouter).
    pub upstream_inference_cost: rust_decimal::Decimal,
    /// Cost charged by the upstream's upstream (e.g., the actual model provider).
    pub upstream_upstream_inference_cost: rust_decimal::Decimal,
}

impl CostDetails {
    /// Returns `true` if any cost is non-zero.
    pub fn any_usage(&self) -> bool {
        self.upstream_inference_cost > rust_decimal::Decimal::ZERO
            || self.upstream_upstream_inference_cost
                > rust_decimal::Decimal::ZERO
    }

    /// Appends cost details from another instance.
    pub fn push(&mut self, other: &CostDetails) {
        self.upstream_inference_cost += other.upstream_inference_cost;
        self.upstream_upstream_inference_cost +=
            other.upstream_upstream_inference_cost;
    }
}
